skl2onnx/operator_converters/nearest_neighbours.py

```diff
@@ -5,498 +5,247 @@
-            weighted_distance_name = scope.get_unique_variable_name(
-                'weighted_distance')
-
-            container.add_node('Equal', [labels_name[i], topk_labels_name],
-                               output_label_name[i],
-                               name=scope.get_unique_operator_name('Equal'))
-            apply_cast(scope, output_label_name[i], output_cast_label_name[i],
-                       container, to=onnx_proto.TensorProto.FLOAT)
-            apply_mul(scope, [output_cast_label_name[i], weights_val],
-                      weighted_distance_name, container, broadcast=0)
-            container.add_node('ReduceSum', weighted_distance_name,
-                               output_label_reduced_name[i], axes=[1],
-                               name=scope.get_unique_operator_name(
-                                   'ReduceSum'))
+    X = operator.inputs[0]
+    op = operator.raw_operator
+    opv = container.target_opset
+    dtype = container.dtype
+
+    if isinstance(X.type, Int64TensorType):
+        X = OnnxCast(X, to=container.proto_dtype, op_version=opv)
+
+    options = container.get_options(op, dict(optim=None))
+
+    single_reg = (not hasattr(op, '_y') or len(op._y.shape) == 1 or
+                  len(op._y.shape) == 2 and op._y.shape[1] == 1)
+    ndim = 1 if single_reg else op._y.shape[1]
+
+    metric = op.effective_metric_
+    neighb = op._fit_X.astype(container.dtype)
+    k = op.n_neighbors
+    training_labels = op._y if hasattr(op, '_y') else None
+    distance_kwargs = {}
+    if metric == 'minkowski':
+        if op.p != 2:
+            distance_kwargs['p'] = op.p
+        else:
+            metric = "euclidean"
+
+    weights = op.weights if hasattr(op, 'weights') else 'distance'
+    if weights == 'uniform':
+        top_indices = onnx_nearest_neighbors_indices(
+            X, neighb, k, metric=metric, dtype=dtype,
+            op_version=opv, optim=options.get('optim', None),
+            **distance_kwargs)
+        top_distances = None
+    elif weights == 'distance':
+        top_indices, top_distances = onnx_nearest_neighbors_indices(
+            X, neighb, k, metric=metric, dtype=dtype,
+            op_version=opv, keep_distances=True,
+            optim=options.get('optim', None),
+            **distance_kwargs)
+    else:
+        raise RuntimeError(
+            "Unable to convert KNeighborsRegressor when weights is callable.")
+
+    shape = OnnxShape(top_indices, op_version=opv)
+    flattened = OnnxFlatten(top_indices, op_version=opv)
+    if training_labels is not None:
+        if ndim > 1:
+            # shape = (ntargets, ) + shape
+            training_labels = training_labels.T
+            shape = OnnxConcat(np.array([ndim], dtype=np.int64),
+                               shape, op_version=opv, axis=0)
+            axis = 2
+        else:
+            training_labels = training_labels.ravel()
+            axis = 1
+
+        if training_labels.dtype == np.int32:
```

Comment:
Where do you handle string labels?

Reply:
Apparently I did not, that also means it is not covered by any unit test. I'll need to add more tests tomorrow.

Reply:
Yeah, we don't have unit tests for that. Also, if you are updating the unit tests, can you make them use fit_classification_model() and fit_regression_model() from tests_helper.py.

Reply:
All fixed.

Reply:
The last failure comes from neighbours. When there neighbours are at the same exact distance, scikit-learn and onnx don't necessarily select the sames.

Reply:
Okay.